kf_lib_data_ingest/validation/reporting/base.py

```diff
@@ -0,0 +1,245 @@
+        req_result_keys = {
+            "type",
+            "description",
+            "is_applicable",
+            "errors",
+            "inputs",
+        }
+
+        assert_safe_type(results["counts"], dict)
+        assert_safe_type(results["files_validated"], list)
+        assert_safe_type(results["validation"], list)
+
+        # Check for top level required keys by attempting to access all
+        # expected keys
+        try:
+            for k in req_keys:
+                results[k]
+        except KeyError:
+            self.logger.error(
+                f"Validation results dict has keys: "
+                f"{pformat(results.keys())} but is missing one or more "
+                f"required keys: {pformat(req_keys)}"
+            )
+            raise
+
+        # Check for validation result dict required keys by attempting to
+        # access all expected keyss
+        try:
+            for r in results["validation"]:
+                for k in req_result_keys:
+                    r[k]
+        except KeyError:
+            self.logger.error(
+                f"Validation test result dict:\n{pformat(r)}"
+                "\n is missing one or more required keys: "
+                f"{pformat(req_result_keys)}"
+            )
+            raise
+
+            # Check types
+            assert_safe_type(r["type"], str)
+            assert_safe_type(r["description"], str)
+            assert_safe_type(r["is_applicable"], bool)
+            assert_safe_type(r["errors"], list, dict)
+
+            # -- Check error dicts --
+            # Ensure error dicts follow expected structure by attempting to
+            # access all keys and values. Catch any exception, log that there
+            # was a validation error in the validation result's error dict,
+            # re-raise the original exception
+            try:
+                errs = r["errors"]
+                if r["type"] == "relationship":
+                    for e in errs:
+                        t, v = e["from"]
+                        if e["to"]:
+                            for to_node in e["to"]:
+                                t1, v1 = to_node
+                        for (t2, v2), files in e["locations"].items():
+                            pass
+                elif r["type"] == "attribute":
+                    assert_all_safe_type(errs.keys(), str)
+                    assert_all_safe_type(errs.values(), set)
+
+                elif r["type"] == "count":
+                    assert_all_safe_type(errs.keys(), str)
+                    assert_all_safe_type(errs.values(), int)
+
+            except Exception:
+                self.logger.error(
+                    "Badly formatted `errors` in validation result dict: "
+                    f"\n{pformat(r)}\n"
+                    f"Please see {SAMPLE_VALIDATION_RESULTS} "
+                    f"for expected validation results format."
+                )
+                raise
```

Comment:
feels weird to have this here and not just in a unit test

Reply:
Yea I guess I could remove it now that the interface is stable 🤷‍♀️. Plus people can look at `sample_validation_results.py` for a look at the result format